A lot more use cases for grouping operator (#21883)
mdn/content: files/en-us/web/javascript/reference/operators/grouping/index.md

{{jsSidebar("Operators")}}

The grouping operator `( )` controls the precedence of evaluation in expressions. It also acts as a container for arbitrary expressions in certain syntactic constructs, where ambiguity or syntax errors would otherwise occur.

{{EmbedInteractiveExample("pages/js/expressions-groupingoperator.html")}}

## Syntax

```js-nolint
(expression)
```

### Parameters

- `expression`

>   - : Any [expression](/en-US/docs/Web/JavaScript/Reference/Operators), including [comma-joined](/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator) ones, to be evaluated.

Reviewer: Couldn't tell you why I prefer it, but maybe
```suggestion
  - : Any [expression](/en-US/docs/Web/JavaScript/Reference/Operators) to be evaluated, including [comma-joined](/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator) expressions.
```

> 

Reviewer 2: It is odd for me to see this "Parameters" form for an operator. With, say, `a ** b` we don't say `a` and `b` are parameters, we say they are operands, right? Why is it different here?

Author: I do notice we don't add `Parameters` for some operator/statement pages. But IMO it makes sense for operators—expressions may throw, so you may end up with an `Exceptions` section; and you almost certainly want to have a `Return value` section. In this case you probably want a `Parameters` section as well, for the sake of structure completeness. We may interpret it as "parameters for the grammar production".

Reviewer 2: Do you think we should have "Parameters" for all the operators, including things like `x ** y`? I would have thought "Operands" would be more conventional.

I mean, I don't desperately mind, I mostly want to be consistent.

Author: Hmm, now I'm slightly torn. I still favor "Parameters" because: (1) consistency with method references (2) we document expressions that are not operators (e.g. dynamic import, function expressions), so "operands" doesn't generalize. If we can propagate the idea that "syntactical structures also have parameters", then it naturally applies to operators, expressions, statements, and everything else.

Reviewer 2: I'm fine with this but think it's worth having a quick check that people are in general as it affects a few pages. I'll mention it in the editorial meeting today.

Reviewer 3: I've given it a read and I like the idea of consistancy here - I have no strong feelings about what the section heading should be - @Josh-Cena your reasoning for _Parameters_ seems good to me 👍

## Description

> The grouping operator consists of a pair of parentheses around an expression to override the normal [operator precedence](/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence), so that operators with lower precedence (as low as the [comma](/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator) operator) can be evaluated before an operator with higher precedence. As it sounds, it groups what's inside of the parentheses.

Reviewer: ```suggestion
The grouping operator consists of a pair of parentheses around an expression that groups the contents.
The operator overrides the normal [operator precedence](/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence), so that operators with lower precedence (as low as the [comma](/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator) operator) can be evaluated before an operator with higher precedence.
```

Author: Much better 👍

```suggestion
The grouping operator consists of a pair of parentheses around an expression that groups the contents. The operator overrides the normal [operator precedence](/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence), so that operators with lower precedence (as low as the [comma](/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator) operator) can be evaluated before an operator with higher precedence.
```

## Examples

### Using the grouping operator

> Overriding multiplication and division first, then addition and subtraction to evaluate addition first.

Reviewer: Maybe 
```suggestion
Evaluating addition and subtraction before multiplication and division.
```

```js
const a = 1;
const b = 2;
const c = 3;

// default precedence
a + b * c; // 7
// evaluated by default like this
a + (b * c); // 7

// now overriding precedence
// addition before multiplication
(a + b) * c; // 9

// which is equivalent to
a * c + b * c; // 9
```

> Notice in these examples that the left-to-right order of evaluation is still preserved. In other words, the order in which the _operators_ evaluate has changed, but the order in which the _operands_ evaluate has not. For example in this code:

Reviewer: ```suggestion
Notice that in these examples the left-to-right order of evaluation is still preserved. In other words, the order in which the _operators_ evaluate has changed, but the order in which the _operands_ evaluate has not. For example in this code:
```

Reviewer: Actually, I think perhaps "something like":

```suggestion
Notice in these examples that the order in which the _operators_ evaluate has changed, but the order in which the _operands_ evaluate has not. For example in this code the functions `a`, `b`, and `c` are evaluated left-to-right (the normal order of evaluation) before the operator order is considered.
```
And then removing the first sentence afterwards.

Ignore if this is "wrong" :-)

Author: Your interpretation is right! :)

```suggestion
Notice in these examples that the order in which the _operators_ evaluate has changed, but the order in which the _operands_ evaluate has not. For example, in this code, the functions `a`, `b`, and `c` are evaluated left-to-right (the normal order of evaluation) before the operator order is considered.
```

> Notice in these examples that the order in which the _operators_ evaluate has changed, but the order in which the _operands_ evaluate has not. For example, in this code, the function invocations `a()`, `b()`, and `c()` are evaluated left-to-right (the normal order of evaluation) before the operator order is considered.

Reviewer: Yes, this all flows well for me now.

```js
a() * (b() + c());
```

The function `a` will be called before the function `b`, which will be called before the function `c`. For more on operator precedence, see its [reference page](/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence).

> ### Using the grouping operator to eliminate ambiguity

Author: ```suggestion
### Using the grouping operator to eliminate parsing ambiguity
```

> An expression statement (a [statement](/en-US/docs/Web/JavaScript/Reference/Statements) consisted of a single expression) cannot start with the keyword `function`, because that would be ambiguous with a [function declaration](/en-US/docs/Web/JavaScript/Reference/Statements/function). This means the following [IIFE](/en-US/docs/Glossary/IIFE) syntax is invalid:

Reviewer: "ambiguous with" does not feel correct. Maybe:

```suggestion
An expression statement (a [statement](/en-US/docs/Web/JavaScript/Reference/Statements) consisted of a single expression) cannot start with the keyword `function`, because that could represent a [function declaration](/en-US/docs/Web/JavaScript/Reference/Statements/function). This means the following [IIFE](/en-US/docs/Glossary/IIFE) syntax is invalid:
```

Author: ```suggestion
An expression statement (a [statement](/en-US/docs/Web/JavaScript/Reference/Statements) consisted of a single expression) cannot start with the keyword `function`, because the parser would see it as the start of a [function declaration](/en-US/docs/Web/JavaScript/Reference/Statements/function). This means the following [IIFE](/en-US/docs/Glossary/IIFE) syntax is invalid:
```

```js example-bad
function () {
  // code
}();
```

The grouping operator can be used to eliminate this ambiguity, since when the parser sees the left parenthesis, it knows that what follows must be an expression instead of a declaration.

```js
(function () {
  // code
})();
```

You may also use the [`void`](/en-US/docs/Web/JavaScript/Reference/Operators/void#immediately_invoked_function_expressions) operator to eliminate ambiguity.

In an [arrow function](/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) concise body (one that directly returns an expression without the keyword `return`), the grouping operator can be used to return an object literal expression, because otherwise the left curly brace would be interpreted as the start of the function body.

```js
const f = () => ({ a: 1 });
```

If a property is accessed on a number literal, the [property accessor](/en-US/docs/Web/JavaScript/Reference/Operators/Property_Accessors) dot `.` may be ambiguous with a decimal point, unless the number already has a decimal point. You can wrap integer literals in parentheses to eliminate this ambiguity.

```js
(1).toString(); // "1"
```

> <!-- TODO in the future we can add a decorator example -->

Author: ```suggestion
<!-- TODO in the future we can add a decorator section -->
```

### Grouping operator and automatic semicolon insertion

The grouping operator can mitigate [automatic semicolon insertion](/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#automatic_semicolon_insertion) (ASI) pitfalls. For example, the `return` keyword and the returned expression cannot have a line break in between:

```js-nolint example-bad
function sum(a, b) {
  return
    a + b;
}
```

> This code will return `undefined`, because a semicolon is inserted directly after the `return` keyword, which causes the function to return immediately without evaluating `a + b`. In case the returned expression is long and you want to keep it well-formatted, you may use the grouping operator to signify that the `return` keyword is followed by an expression and prevent semi-colon insertion:

Author: ```suggestion
This code will return `undefined`, because a semicolon is inserted directly after the `return` keyword, which causes the function to return immediately without evaluating `a + b`. In case the returned expression is long and you want to keep it well-formatted, you may use the grouping operator to signify that the `return` keyword is followed by an expression and prevent semicolon insertion:
```
